KAFKA-9564: Local Tiered Storage implementation for Remote Storage Manager

kamalcph opened the pull request:
Tiered Storage implementation for Remote Storage Manager using the local file system. This will be used for system integration tests as well in the followup PRs. 

Primary Contributor: Alexandre Dupriez <alexandre.dupriez@gmail.com>

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

kamalcph committed:
Local Tiered Storage implementation for Remote Storage Manager.

kamalcph commented:
@satishd 
Can you please take a look when you get chance?

divijvaidya commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorage.java:
```diff
@@ -0,0 +1,564 @@
+
+            if (!existed) {
+                logger.info("Creating directory: [{}]", storageDirectory.getAbsolutePath());
+                storageDirectory.mkdirs();
```
please use `Files.createDirectories`. The new (relatively) Files API introduced in JDK 7 has better error handling semantics.

showuon commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorage.java:
```diff
@@ -0,0 +1,564 @@
+
+    /**
+     * Used to notify users of this storage of internal updates - new topic-partition recorded (upon directory
+     * creation) and segment file written (upon segment file write(2)).
```
What does `write(2)` mean?

showuon commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorage.java:
```diff
@@ -0,0 +1,564 @@
+        if (transfererClass != null) {
+            try {
+                transferer = (Transferer) getClass().getClassLoader().loadClass(transfererClass).newInstance();
+
```
nit: extra line

kamalcph replied:
Will remove the extra line from all the files once the patch is in good shape for smooth review.

showuon commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorage.java:
```diff
@@ -0,0 +1,564 @@
+
+        } else {
+            storageDirectory = new File(storageDir + "/" + ROOT_STORAGES_DIR_NAME);
+            // NOTE: Provide the relative storage directory path to view the files in the same directory when running tests.
+            // storageDirectory = new File(new File("."), ROOT_STORAGES_DIR_NAME + "/" + storageDir);
```
What does these 2 lines mean?

kamalcph replied:
LocalTieredStorage implementation can be used in two modes: 

1. Integration test and
2. To bring Kafka in the local machine with `LocalTieredStorage` implementation as RSM for quick testing by adding `kafka-storage-<version>-test.jar` in the classpath, then enabling the `INCLUDE_TEST_JARS` flag before running the server

In mode 2, if the commented-out line#273 is enabled, then it creates the `storageDirectory` from the present kafka working directory which can cause confusion:

(eg)
Kafka is installed on `/opt/kafka` and `storageDirectory` is configured as `/tmp/kafka-tiered-storage`, then it will create it the directory in `/opt/kafka/kafka-tiered-storage/tmp/kafka-tiered-storage` instead of `/tmp/kafka-tiered-storage`. 

For the integration test, the `storageDirectory` won't be set so it uses the java temporary directory.

showuon replied:
Hmm.. should we keep it here? Or should we add more context in the comment? Obviously the comment is confusing now.

kamalcph replied:
Added the context as a comment on the code.

showuon commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorage.java:
```diff
@@ -0,0 +1,564 @@
+            try {
+                fileset = openFileset(storageDirectory, id);
+
+                logger.info("Offloading log segment for {} from offset={}", id.topicIdPartition(), data.logSegment());
```
I don't think this line will output the expected result. I think we should log (1) partition (2) segment (3) offset

kamalcph replied:
Sample Log: We can improve this log statement by extracting the offset from the filename.

```
Offloading log segment for topicA-0 from offset=/var/folders/_y/m0wdyqps33sczpp2_3t_6nk00000gn/T/kafka-1852885522885661832/topicA-0/00000000000000000000.log 
```

showuon replied:
Thanks for the sample log output.
So, maybe we can log as: 
`"Offloading log segment for {} from segment={}"`

showuon commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorage.java:
```diff
@@ -0,0 +1,564 @@
+        });
+    }
+
+    // New API to be added in RemoteStorageManager
+    // @Override
```
Should be removed? Or remove the `//` in front of `@Override`?

kamalcph replied:
RemoteStorageManager does not have the `deletePartition` API. Once added, will remove this comment. This API is required to delete the partition in one shot instead of removing the segments one by one.

showuon replied:
I see. Do we have JIRA ticket tracking it?

kamalcph replied:
Filed: https://issues.apache.org/jira/browse/KAFKA-15166 to track it.

showuon reviewed (commented):
Had a reviewed till `LocalTieredStorageEvent.java`. Left some comments.

showuon commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorageEvent.java:
```diff
@@ -0,0 +1,187 @@
+/**
+ * Represents an interaction between a broker and a second-tier storage. This type of event is generated
+ * by the {@link LocalTieredStorage} which is an implementation of the {@link RemoteStorageManager}
+ * operating in Kafka's runtime as the interface between Kafka and external storage systems, through
+ * which all such interactions go through.
```
Sorry, the last sentence is hard to understand: `through which all such interactions go through.` 
Is there better way to describe it?

showuon commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorageEvent.java:
```diff
@@ -0,0 +1,187 @@
+    public int compareTo(LocalTieredStorageEvent other) {
+        requireNonNull(other);
+
+        if (other.timestamp > timestamp) {
+            return -1;
+        }
+        if (other.timestamp < timestamp) {
+            return 1;
+        }
+        return 0;
```
We can directly return 
```
return timestamp - other.timestamp;
```
> Compares this object with the specified object for order. Returns a negative integer, zero, or a positive integer as this object is less than, equal to, or greater than the specified object.

ref: https://docs.oracle.com/javase/8/docs/api/java/lang/Comparable.html#compareTo-T-

divijvaidya commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/RemoteTopicPartitionDirectory.java:
```diff
@@ -0,0 +1,169 @@
+ */
+public final class RemoteTopicPartitionDirectory {
+    private static final Logger LOGGER = getLogger(RemoteTopicPartitionDirectory.class);
+    private static final String UUID_LEGAL_CHARS = "[a-zA-Z0-9_-]{22}";
```
`_` (underscore) is not a valid UUID char as per https://www.ietf.org/rfc/rfc4122.txt

kamalcph replied:
`_` is a valid character in [Base64URL](https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/util/Base64.java#L240). Uuid is encoded to Base64 in the [Uuid#toString](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/Uuid.java#L110) method.

divijvaidya replied:
Interesting. #TIL

divijvaidya commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/RemoteTopicPartitionDirectory.java:
```diff
@@ -0,0 +1,169 @@
+ *                     .                   .  tvHCaSDsQZWsjr5rbtCjxA-time_index
+ *                     .
+ *                     / 5fEBmixCR5-dMntYSLIr1g-3-topic / BFyXlC8ySMm-Uzxw5lZSMg-segment
+ *                                                      . BFyXlC8ySMm-Uzxw5lZSMg-offset_index
```
Can we use `UnifiedLog.IndexFileSuffix` (similar for offset and trx index) here which will use `.` instead of underscore to name the files? This will bring consistency with how we store the file names on log and RemoteIndexCache.

divijvaidya commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorage.java:
```diff
@@ -0,0 +1,564 @@
+    /**
+     * The ID of the broker which owns this instance of {@link LocalTieredStorage}.
+     */
+    public static final String BROKER_ID = "broker.id";
```
Use `KafkaConfig.BrokerIdProp`?

kamalcph replied:
This will make a direct dependency on the `core` module.

divijvaidya replied:
ok

divijvaidya commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorage.java:
```diff
@@ -0,0 +1,564 @@
+
+
+        if (storageDir == null) {
+            storageDirectory = TestUtils.tempDirectory(ROOT_STORAGES_DIR_NAME + "-");
```
This will be deleted on exit because TestUtils.tempDirectory registers `file.deleteOnExit()`. Is this the behaviour we want when we pass deleteOnClose=false?

kamalcph replied:
The assumption made here is `deleteOnClose` flag is applicable only when `storageDir` is set.

divijvaidya commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorage.java:
```diff
@@ -0,0 +1,564 @@
+                    storageDirectory.getAbsolutePath());
+
+        } else {
+            storageDirectory = new File(storageDir + "/" + ROOT_STORAGES_DIR_NAME);
```
You don't need to specify the "/" as it can be OS dependent. File API takes care of that for you with File(parent,child) API: https://docs.oracle.com/javase/8/docs/api/java/io/File.html#File-java.io.File-java.lang.String-

divijvaidya commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorage.java:
```diff
@@ -0,0 +1,564 @@
+            storageDirectory = new File(storageDir + "/" + ROOT_STORAGES_DIR_NAME);
+            // NOTE: Provide the relative storage directory path to view the files in the same directory when running tests.
+            // storageDirectory = new File(new File("."), ROOT_STORAGES_DIR_NAME + "/" + storageDir);
+            final boolean existed = storageDirectory.exists();
```
please use `Files.exists()` which was introduced in JDK 7 in NIO2 APIs. It has better error handling semantics.

divijvaidya commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/RemoteLogSegmentFileset.java:
```diff
@@ -0,0 +1,262 @@
+        TIME_INDEX(false),
+        TRANSACTION_INDEX(true),
+        LEADER_EPOCH_CHECKPOINT(false),
+        PRODUCER_SNAPSHOT(true);
```
Current code expects producer snapshot to be present (although this is incorrect because segments uploaded prior to 2.8 won't have this), hence, this is a mandatory entry. The code fails at:
https://github.com/apache/kafka/blob/43574beb972d47d696e0de077f453b36ce148026/core/src/main/java/kafka/log/remote/RemoteLogManager.java#L592 with null pointer exception if producer snapshot is not available.

kamalcph replied:
Good catch! Updated the patch to make `producerSnapshot` file non-optional. We can discuss further separately whether to make it optional for a smoother upgrade from older server versions (<2.8).

showuon replied:
Good point!

divijvaidya commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/RemoteLogSegmentFileset.java:
```diff
@@ -0,0 +1,262 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.server.log.remote.storage;
```
is this the right place to place these files? Asking because all other TS related files are in `kafka.log.remote`

kamalcph replied:
`RemoteStorageManager` interface is in the `org.apache.kafka.server.log.remote.storage` package. As, `LocalTieredStorage` is one of the implementation of `RemoteStorageManger`, kept them in this package.

divijvaidya replied:
ok

divijvaidya commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/RemoteLogSegmentFileset.java:
```diff
@@ -0,0 +1,262 @@
+    public static RemoteLogSegmentFileset openExistingFileset(final RemoteTopicPartitionDirectory tpDirectory,
+                                                              final Uuid uuid) {
+        final Map<RemoteLogSegmentFileType, File> files =
+                stream(tpDirectory.getDirectory().listFiles())
```
Please use newer API `Files.list()`

divijvaidya replied:
also filter for non-directories

divijvaidya commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/RemoteLogSegmentFileset.java:
```diff
@@ -0,0 +1,262 @@
+            if (!file.exists()) {
+                return true;
+            }
+            return file.delete();
```
You can use `Utils.delete(File)` here

kamalcph replied:
`Utils.delete(File)` doesn't return the deletion status.

divijvaidya replied:
ok

divijvaidya commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/Transferer.java:
```diff
@@ -0,0 +1,28 @@
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+public interface Transferer {
```
a java doc on the purpose of this interface will be very useful here

kamalcph replied:
Added the Javadoc for this interface.

divijvaidya commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorageTest.java:
```diff
@@ -0,0 +1,675 @@
+            final String uuid = id.id().toString();
+
+            return Arrays.asList(
+                    Paths.get(rootPath, topicPartitionSubpath, uuid + "-segment"),
```
these suffix should be moved to a constant

divijvaidya reviewed (commented):
Did a first round. Will continue in upcoming days.

It would be very useful if this PR can add at least one test utilizing this setup.

showuon commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorageHistory.java:
```diff
@@ -0,0 +1,115 @@
+     * Returns the list of events accumulated by this instance of history from the {@link LocalTieredStorage}
+     * it captures events from.
+     *
+     * @param type The type of the events to retrieve ((e.g. offload or fetch a segment, fetch a time index, etc.)
```
nit: double `((`

showuon commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorageHistory.java:
```diff
@@ -0,0 +1,115 @@
+    /**
+     * Returns the latest event captured so far of the given type and relating to the given topic-partition.
+     *
+     * @param type The type of the events to retrieve ((e.g. offload or fetch a segment, fetch a time index, etc.)
```
nit: double `((`

showuon commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorageListener.java:
```diff
@@ -0,0 +1,74 @@
+            for (final LocalTieredStorageListener listener: listeners) {
+                try {
+                    listener.onStorageEvent(event);
+
```
nit: extra line

kamalcph replied:
Will remove the extra line from all the files once the patch is in good shape for smooth review.

showuon commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorageListener.java:
```diff
@@ -0,0 +1,74 @@
+                    listener.onStorageEvent(event);
+
+                } catch (Exception e) {
+                    LOGGER.error("Caught failure from listener", e);
```
Should we log which listener it is?

kamalcph committed:
Addressed the review comments to use consistent file suffixes.

- Used `Files` API wherever it's applicable.

kamalcph commented:
> Did a first round. Will continue in upcoming days.
> 
> It would be very useful if this PR can add at least one test utilizing this setup.

Thanks, @divijvaidya and @showuon for reviewing the PR! Addressed your review comments. PTAL. I'll open a separate PR to add the `TieredStorageTestHarness` classes to effectively write an integration test.

showuon commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorageTest.java:
```diff
@@ -0,0 +1,677 @@
+         * @param id The unique ID of the remote log segment and associated resources (e.g. offset and time indexes).
+         * @param segment The segment stored on Kafka's local storage.
+         */
+        public void verifyLogSegmentFilesAbsent(final RemoteLogSegmentId id, final LogSegmentData segment) {
+            expectedPaths(id).forEach(this::assertFileDoesNotExist);
```
Why do we need the 2nd parameter `segment` here?

showuon commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorageTest.java:
```diff
@@ -0,0 +1,677 @@
+        }
+
+        /**
+         * Compare the content of the remote segment with the provided {@code data} array.
+         * This method does not fetch from the remote storage.
```
Don't know where `data` come from?

showuon commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorageTest.java:
```diff
@@ -0,0 +1,677 @@
+        }
+
+        private void assertFileExists(final Path path) {
+            if (!path.toFile().exists()) {
+                throw new AssertionError(format("File %s does not exist", path));
```
nit: `assertTrue(path.toFile().exists(), format("File %s does not exist", path));`

showuon commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorageTest.java:
```diff
@@ -0,0 +1,677 @@
+        }
+
+        private void assertFileDoesNotExist(final Path path) {
+            if (path.toFile().exists()) {
+                throw new AssertionError(format("File %s should not exist", path));
```
nit: `assertFalse(path.toFile().exists(), format("File %s does not exist", path));`

showuon reviewed (commented):
Thanks for the update. Left some more comments.

kamalcph committed:
Addressed the review-2 comments

kamalcph commented:
@showuon @divijvaidya 
Addressed your review comments. Please take another look. Thanks!

jeqo commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorage.java:
```diff
@@ -0,0 +1,578 @@
+    /**
+     * The root directory of this storage.
+     */
+    public static final String STORAGE_DIR_PROP = "dir";
```
Not specifically to this PR: 
Just realized that Tiered Storage uses a different suffix than most Kafka code base (`_PROP` instead of `_CONFIG`).
Even though it's a minor issue, as config names are public and potentially used in plugins: would it make sense to address this early?

jeqo commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorage.java:
```diff
@@ -0,0 +1,578 @@
+     * this storage. The only reason the "transferer" abstraction exists is to be able to simulate
+     * file copy errors and exercise the associated failure modes.
+     */
+    public static final String TRANSFERER_CLASS_PROP = "transferer";
```
```suggestion
    public static final String TRANSFERER_CLASS_PROP = "transferer.class";
```

jeqo commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorage.java:
```diff
@@ -0,0 +1,578 @@
+     */
+    public static final String BROKER_ID = "broker.id";
+
+    private static final String ROOT_STORAGES_DIR_NAME = "kafka-tiered-storage";
```
nit:
```suggestion
    private static final String ROOT_STORAGE_DIR_NAME = "kafka-tiered-storage";
```

jeqo commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorage.java:
```diff
@@ -0,0 +1,578 @@
+
+    private volatile int brokerId = -1;
+
+    private volatile Logger logger = LoggerFactory.getLogger(LocalTieredStorage.class);
```
Curious about Logger being volatile. If needed, could we comment why?

jeqo commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorage.java:
```diff
@@ -0,0 +1,578 @@
+ * Given the root directory of the storage, segments and associated files are organized as represented below.
+ * </p>
+ * <code>
+ * / storage-directory  / LWgrMmVrT0a__7a4SasuPA-0-topic / bCqX9U--S-6U8XUM9II25Q.log
```
Is there any reason to do not prefix by topic-partition, and then by offset to make the resulting structure easier to navigate?

kamalcph replied:
Topics with the same name can be recreated with different Uuid. Currently, the RemoteStorageManger exposes API only to delete the segments, not the partition. When a topic gets deleted, there will be empty partition directories. To differentiate the stale topic directory from the active one, we need Uuid in the directory name.

Renaming the directory to `topic-partition-uuid`, requires a good amount of change. We can do it in the follow-up PR if required.

jeqo replied:
> To differentiate the stale topic directory from the active one, we need Uuid in the directory name.

Sure, I wasn't advocating to have a partition directory, but something like this: 
```suggestion
 * / storage-directory  / topic-0-LWgrMmVrT0a__7a4SasuPA / 0000000000-bCqX9U--S-6U8XUM9II25Q.log
```

I agree changing the file structure may take a bit more, but the directory should be doable. I just added some suggestions on the 3 places that need changes if we want to update the layout with:

> In case we want to adopt the new structure:

on the comment. Let me know if this may be enough to improve the structure.

kamalcph replied:
Updated the directory structure to `topic-partition-uuid`.

jeqo commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorageEvent.java:
```diff
@@ -0,0 +1,180 @@
+     * The nature of the interaction.
+     */
+    public enum EventType {
+        OFFLOAD_SEGMENT,
```
```suggestion
        COPY_SEGMENT,
```

jeqo commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorage.java:
```diff
@@ -0,0 +1,578 @@
+    @Override
+    public InputStream fetchLogSegment(final RemoteLogSegmentMetadata metadata,
+                                       final int startPos) throws RemoteStorageException {
+        return fetchLogSegment(metadata, startPos, Integer.MAX_VALUE);
```
Maybe use `metadata.segmentSizeInBytes()` instead of max value?

kamalcph replied:
`endPos` is currently unused in the LocalRemoteStorage. We can refactor it when we are using the value.

The HDFS implementation internally checks for MAX_VALUE and computes the file length:

```
long actualFileLength = fs.getFileStatus(dataPath).getLen();
fileLen = (dataPosition.getLength() == Integer.MAX_VALUE) ?
  | Math.min(endPos, actualFileLength) :
  | Math.min(endPos, dataPosition.getPos() + dataPosition.getLength());
```

jeqo replied:
Agree, let's make a note to remind us

kamalcph replied:
Updated.

jeqo commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorage.java:
```diff
@@ -0,0 +1,578 @@
+        checkArgument(startPos >= 0, "Start position must be positive", startPos);
+        checkArgument(endPos >= startPos,
+                "End position cannot be less than startPosition", startPos, endPos);
+
```
maybe also validate `metadata.segmentSizeInBytes >= endPos`?

kamalcph replied:
`endPos` is currently unused in the LocalRemoteStorage. We can refactor it when we are using the value.

kamalcph replied:
Updated.

jeqo commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorage.java:
```diff
@@ -0,0 +1,578 @@
+
+                File file = fileset.getFile(fileType);
+                final InputStream inputStream = (fileType.isOptional() && !file.exists()) ?
+                        new ByteArrayInputStream(new byte[0]) : newInputStream(file.toPath(), READ);
```
Based on latest [RSM docs](https://github.com/apache/kafka/pull/13923) files not found, like transactional index that is optional, should return `RemoteResourceNotFound`

kamalcph replied:
Will fix this as part of https://issues.apache.org/jira/browse/KAFKA-14993 ticket

jeqo commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorage.java:
```diff
@@ -0,0 +1,578 @@
+
+    private static final String ROOT_STORAGES_DIR_NAME = "kafka-tiered-storage";
+
+    private volatile File storageDirectory;
```
any reason to not using Path instead of File?

kamalcph replied:
No strong preference, to avoid wrapping the IOException as RuntimeException. We can refactor this in the subsequent PR if required.

jeqo commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorage.java:
```diff
@@ -0,0 +1,578 @@
+            return;
+        }
+
+        Arrays.stream(files)
```
Maybe use `Files.list(storageDirPath)`?

kamalcph replied:
`Files.list` throws IOException and it needs to be wrapped with RuntimeException to avoid bubbling up the exception to the callers.

jeqo reviewed (commented):
Thanks @kamalcph, this PR will be very useful to start ensuring TS is fully functional. 
Took an initial review on the LocalTS.

clolov commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorageHistory.java:
```diff
@@ -0,0 +1,115 @@
+ * {@link LocalTieredStorageHistory#listenTo(LocalTieredStorage)} is called.
+ */
+/* @ThreadSafe */
+public final class LocalTieredStorageHistory {
```
This history is only in-memory, unless I am reading the code wrongly. Are there plans to write it somewhere for debugging? I suspect the main reason behind it is for assertions in tests via the LocalTieredStorageCondition.

kamalcph replied:
yes, you're right. History is maintained only in memory and to assert the interaction with the `RemoteStorageManager`. 

No plans made so far to make it durable. This framework is mainly used to assert the functionality end-to-end instead of stress-test where there can be a huge number of events for a single test.

clolov reviewed (commented):
Great job, overall this makes sense to me, just the question for the introduction of the history!

divijvaidya commented:
@kamalcph Please don't wait for my review on this. I think multiple folks are already on top of this.

kamalcph committed:
Addressed the review-3 comments.

showuon commented:
Thanks for the update. Will have a 2nd round of review in the following days.

jeqo commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorageTest.java:
```diff
@@ -0,0 +1,673 @@
+    private static final DateTimeFormatter DATE_TIME_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd-HH:mm:ss");
+
+    private String generateStorageId(String testName) {
+        return format("%s-%s-%s",
```
Probably add to the directory already mapped on .gitignore
```suggestion
        return format("kafka-tiered-storage/%s-%s-%s",
```

jeqo commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorageTest.java:
```diff
@@ -0,0 +1,673 @@
+            final String rootPath = getStorageRootDirectory();
+            TopicPartition tp = topicIdPartition.topicPartition();
+            final String topicPartitionSubpath = format("%s-%d-%s", topicIdPartition.topicId(),
+                   tp.partition(), tp.topic());
```
In case we want to adopt the new structure: 

```suggestion
            final String topicPartitionSubpath = format("%s-%d-%s",
                    tp.topic(), tp.partition(), topicIdPartition.topicId());
```

jeqo commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/RemoteTopicPartitionDirectory.java:
```diff
@@ -0,0 +1,169 @@
+    private static String toString(TopicIdPartition topicIdPartition) {
+        Uuid uuid = topicIdPartition.topicId();
+        TopicPartition tp = topicIdPartition.topicPartition();
+        return uuid.toString() + "-" + tp.partition() + "-" + tp.topic();
```
In case we want to adopt the new structure:
```suggestion
        return tp.topic() + "-" + tp.partition() + "-" + uuid.toString();
```

jeqo commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/RemoteTopicPartitionDirectory.java:
```diff
@@ -0,0 +1,169 @@
+    /**
+     * The format of a Kafka topic-partition directory. Follows the structure Uuid-partition-Topic.
+     */
+    private static final Pattern FILENAME_FORMAT = compile("(" + UUID_LEGAL_CHARS + ")-(\\d+)-(" + Topic.LEGAL_CHARS + "+)");
+    private static final int GROUP_UUID = 1;
+    private static final int GROUP_PARTITION = 2;
+    private static final int GROUP_TOPIC = 3;
```
In case we want to adopt the new structure:
```suggestion
    private static final Pattern FILENAME_FORMAT = compile("(" + Topic.LEGAL_CHARS + "+)-(\\d+)-(" + UUID_LEGAL_CHARS + ")");
    private static final int GROUP_UUID = 3;
    private static final int GROUP_PARTITION = 2;
    private static final int GROUP_TOPIC = 1;
```

showuon commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/Transferer.java:
```diff
@@ -0,0 +1,33 @@
+
+/**
+ * The implementation of the transfer of the data of the canonical segment and index files to
+ * this storage. The only reason the "transferer" abstraction exists is to be able to simulate
+ * file copy errors and exercise the associated failure modes.
```
I didn't see the `simulate file copy errors` in this PR, which should be expected, right? Those will appear in the real test cases, right?

kamalcph replied:
yes, correct. The `transferer` can be overriden in the respective test to simulate the IO errors.

showuon commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorage.java:
```diff
@@ -0,0 +1,578 @@
+                storageListeners.onStorageEvent(eventBuilder.withFileset(fileset).build());
+
+            } catch (final Exception e) {
+                //
```
nit: additional comment line

showuon commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorage.java:
```diff
@@ -0,0 +1,578 @@
+                // associated offset and time indexes stored as well. Here, delete any file which was copied
+                // before the exception was hit. The exception is re-thrown as no remediation is expected in
+                // the current scope.
+                //
```
ditto

showuon commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorage.java:
```diff
@@ -0,0 +1,578 @@
+            if (notADirectory.isPresent()) {
+                logger.warn("Found file [{}] which is not a remote topic-partition directory. " +
+                        "Stopping the deletion process.", notADirectory.get());
+                //
```
ditto

showuon commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorage.java:
```diff
@@ -0,0 +1,578 @@
+                // If an unexpected state is encountered, do not proceed with the delete of the local storage,
+                // keeping it for post-mortem analysis. Do not throw either, in an attempt to keep the close()
+                // method quiet.
+                //
```
ditto

showuon commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorageCondition.java:
```diff
@@ -0,0 +1,220 @@
+     */
+    public final LocalTieredStorageCondition and(final LocalTieredStorageCondition conjuct) {
+        //
+        // To keep things simple, only authorize append to the condition chain of elementary (not composed)
```
nit: only authorize [to] append to ...

showuon commented on storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorageTest.java:
```diff
@@ -0,0 +1,673 @@
+         * @param id The unique ID of the remote log segment and associated resources (e.g. offset and time indexes).
+         * @param segment The segment stored on Kafka's local storage.
+         */
+        public void verifyContainsLogSegmentFiles(final RemoteLogSegmentId id, final LogSegmentData segment) {
```
nit: unused 2nd parameter

showuon reviewed (commented):
Had a 2nd pass, LGTM! Just some minor comments. Thanks for the PR!

kamalcph committed:
Update the local-log partition directory to topic-partition-uuid format.

kamalcph committed:
Addressed the review-4 comments.

showuon reviewed (approved):
LGTM! Thanks for the PR!

kamalcph committed:
Updated the javadoc.

kamalcph commented:
@jeqo @showuon @satishd 

Addressed your review comments. PTAL. We can work on appending the offset as prefix to the remote log file-set in the next framework PR. Thanks!

jeqo reviewed (approved):
LGTM, thanks @kamalcph !

satishd commented:
> We can work on appending the offset as prefix to the remote log file-set in the next framework PR. Thanks!
@kamalcph Sure, this can be done in other PR. Please file a JIRA if you have not yet so that we will not miss it.

satishd reviewed (approved):
Thanks @kamalcph for addressing the review comments. LGTM.

kamalcph commented:
> > We can work on appending the offset as prefix to the remote log file-set in the next framework PR. Thanks!
> > @kamalcph Sure, this can be done in other PR. Please file a JIRA if you have not yet so that we will not miss it.

Filed https://issues.apache.org/jira/browse/KAFKA-15194 to track this separately.

satishd commented:
A few tests failed in jenkins jobs which are not related to the change, merging to trunk.

satishd commented:
Sorry for missing to add the PR description in the commit message. 

```
Tiered Storage implementation for Remote Storage Manager using the local file system. This will be used for system integration tests as well in the followup PRs.

Primary Contributors: Alexandre Dupriez <alexandre.dupriez@gmail.com>, Kamal Chandraprakash<[kamal.chandraprakash@gmail.com](mailto:kamal.chandraprakash@gmail.com)>

```

divijvaidya commented:
Hey @satishd 
In case you didn't know, you can add co-authors in the commit by adding a line at the end of commit such as:
```
---------

Co-authored-by: Person2 <person2@email.com>
```

GitHub automatically recognizes it and attributes the commit to both, e.g. check this commit https://github.com/apache/kafka/commit/6afcfba9f3949cc8e8eecac42dabb099ba9cc0a9
